### PR TITLE
Cache graphql in useGalleryItem

### LINKED ISSUE #6027: Performance issue on GalleryItem
### What happened?

If you take a look on browser console you can see that `nftById` query is called 5 times.

### Step to reproducs

- any chain
- go to gallery item detail

### Why?

Caused by `useGalleryItem()` composable which is called by 5 different components

### Expected Behavior

- run the query only one time
- (store current item result in store => history mode vuex/pinia)
- share data between component (easy solution use props)

Lmk If there is a better solution

### PR DESCRIPTION
**Pull request by daiagi**
**Thank you for your contribution** to the [KodaDot - One Stop Shop for Polkadot NFTs](https://kodadot.xyz).

👇 \_\_ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [x] Feature
- [x] Refactoring

## Context

- [x] Closes #6027 
- [ ] Requires deployment <snek/rubick/worker>

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality

#### Optional

tested on ksm/gallery/17977877-e63e926f7db5997e5c-STGR-109_ASCETA-00000109

- [ ] I've tested it at </ksm/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/ksm/transfer/?target=1cAsKZYNRb8dkSCpn4eVkEn6ycNZTGoDo5dGDgB8J1UUWK8)

#### Community participation

- [ ] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

Before: (Beta)
image

After:
image

### <samp>🤖 Generated by Copilot at 497d4f5</samp>

This pull request enhances the gallery item component by using the `useQuery` hook from the `@tanstack/vue-query` library to fetch and cache data from GraphQL. It also adds the necessary dependencies, plugins, and composables to support this feature.

### <samp>🤖 Generated by Copilot at 497d4f5</samp>

> _Sing, O Muse, of the clever pull request_
> _That brought to the Nuxt app a new zest_
> _With `useCachedGraphql`, the composable divine_
> _That fetches and caches the data sublime_

### CODE REVIEW

File: plugins/vue-query-plugin.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import { QueryClient, VueQueryPlugin } from '@tanstack/vue-query'
+import Vue from 'vue'
+
+const queryClient = new QueryClient()
+Vue.use(VueQueryPlugin, { queryClient })
```
Comment (Jarsen136) [inline plugins/vue-query-plugin.ts:5]: I wonder if we could use `nuxtApp.vueApp.use` instead. of this

File: composables/useCachedGraphql.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import resolveQueryPath from '@/utils/queryPathResolver'
+import { useQuery } from '@tanstack/vue-query'
+
+const useQueryParams = ({ queryPrefix, clientName }) => {
```
[the remark this answers is not in the document]
Comment (Jarsen136) [reply composables/useCachedGraphql.ts:4]: We could reuse this function.

File: plugins/vue-query-plugin.ts (hunk 1 above)
Comment (daiagi) [reply plugins/vue-query-plugin.ts]: done in https://github.com/kodadot/nft-gallery/pull/6053/commits/93f2f949582e35afd0160f0589423c5d70dbe838

**Comment by daiagi** [comment]
> We could reuse this function.

done in https://github.com/kodadot/nft-gallery/pull/6053/commits/98cd8a6a295a9ed082ded636afe3b172d4ba782c

File: nuxt.config.js [hunk 3]
```diff
@@ -164,6 +164,7 @@ export default defineNuxtConfig({
     '~/plugins/vueClipboard',
     '~/plugins/vueSocialSharing',
     '~/plugins/vueTippy',
+    '~/plugins/vue-query-plugin',
```
Comment (roiLeo) [inline nuxt.config.js:167]: Nuxt 3 auto imported plugin, I would rename file to `vue-query.client.ts`

```suggestion
```

**Review by roiLeo — APPROVED** [review]
Not the optimal solution I was looking for since it include a new package but if it's working like excepted lets go.

File: nuxt.config.js (hunk 3 above)
Comment (daiagi) [reply nuxt.config.js:167]: i did try to remove the line and it didn't work.
i'll try again with your name suggestion

Comment (daiagi) [reply nuxt.config.js]: yeah... not working with that line :shrug:

Comment (roiLeo) [reply nuxt.config.js]: Alright, I can't wait to migrate from nuxt-bridge...

**Review by Jarsen136 — APPROVED** [review]

**Comment by prury** [comment]
My numbers:
<table role="table">
<thead>
<tr>
      <th>Before:</th>
      <th>After:</th>
</tr>
</thead>
<tbody>
<tr>
     <td> </td>
     <td>
</td>
   </td>
</tr>
</tbody>
</table>

**Comment by preschian** [comment]
> Edit(NftbyID):

@prury aw, seems worse. is that the first request on the `after` screenshot? because after some seconds we have a `subscription` request that will always refetch the `nfyById` again

**Comment by preschian** [comment]
interesting, this is what I got. compared side by side

https://github.com/kodadot/nft-gallery/assets/734428/330eedb3-2fcd-4862-bdea-a5bb492d996b

@daiagi, could you check what if move `useSubscriptionGraphql()` section only? maybe we don't need a new package at all

**Comment by prury** [comment]
> Edit(NftbyID):
> 
> @prury aw, seems worse. is that the first request on the `after` screenshot? because after some seconds we have a `subscription` request that will always refetch the `nfyById` again

yes, but i saw you got similar results now :)

**Comment by daiagi** [comment]
@prury @preschian 
the lib I've added does a re fetch when page get focus again/ click in dev tools and back in app
that's why you may be seeing different results

having said that preschian is right and root reason for multiple fetch seems to be ```useSubscriptionGraphql`` and the addition of this new lib may not be required

changing to draft

**Comment by daiagi** [comment]
Changes:

1. remove the new dependency
2. instead, pass prop to ```GalleryItem``` internal components

###outcome

NftById is called twice (previous 5). one of these is due to  ```useSubscriptionGraphql```


### Room for improvement

make ```useSubscriptionGraphql``` call re fetch only if data changed

File: components/gallery/useGalleryItem.ts [hunk 4]
```diff
@@ -48,7 +58,6 @@ export const useGalleryItem = (nftId?: string) => {
   const queryPath = {
     rmrk: 'chain-rmrk',
     ksm: 'chain-ksm',
-    stmn: 'chain-stmn',
```
Comment (roiLeo) [inline components/gallery/useGalleryItem.ts:51]: remove? merge conflicts?

**Review by roiLeo — APPROVED** [review]
otherwise code lgtm

File: components/gallery/useGalleryItem.ts (hunk 4 above)
Comment (daiagi) [reply components/gallery/useGalleryItem.ts]: restored

**Review by preschian — APPROVED** [review]

**Comment by daiagi** [comment]
@prury you can verify that the amount of times NftById is called is reduced from 5 to 2

And that nothing broke in GalleryItem page

That's all

**Comment by prury** [comment]
> @prury you can verify that the amount of times NftById is called is reduced from 5 to 2
> 
> And that nothing broke in GalleryItem page
> 
> That's all

Gallery Item page is looking good and amount of NftbyID calls reduced, ms slightly higher as we talked before: 

https://github.com/kodadot/nft-gallery/assets/36627808/36d4901e-334b-41a4-99e3-e7e3e6e6a41c

**Comment by yangwao** [comment]
pay 50 usd

**Comment by daiagi** [comment]
@yangwao bot didn't trigger

**Comment by yangwao** [comment]
pay 70 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $70 @ 5.02 USD/DOT ~ 13.944 $DOT
🧗 [1cAsKZYNRb8dkSCpn4eVkEn6ycNZTGoDo5dGDgB8J1UUWK8](https://polkadot.subscan.io/account/1cAsKZYNRb8dkSCpn4eVkEn6ycNZTGoDo5dGDgB8J1UUWK8)
🔗 [0x3a7633ca495ce68346b7281eefb8a90d15650dee665523fc857b53ec8a7896d0](https://polkadot.subscan.io/extrinsic/0x3a7633ca495ce68346b7281eefb8a90d15650dee665523fc857b53ec8a7896d0)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**